File: packages/python/src/tbd_agents/resources/knowledge_items.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any, BinaryIO, Mapping

from tbd_agents._utils import parse_content_disposition_filename
from tbd_agents.types import (
    DownloadedContent,
    KnowledgeItem,
    KnowledgeItemCreate,
    KnowledgeItemUpdate,
    KnowledgeQueryResponse,
)

from .base import BaseResource


class KnowledgeItemsResource(BaseResource):
    def create(self, payload: KnowledgeItemCreate | Mapping[str, Any]) -> KnowledgeItem:
        return self._post("/knowledge-items", payload, model=KnowledgeItem)

    def list(
        self,
        *,
        source_id: str | None = None,
        tags: list[str] | None = None,
        content_type: str | None = None,
    ) -> list[KnowledgeItem]:
        params: dict[str, Any] = {}
        if source_id is not None:
            params["source_id"] = source_id
        if tags:
            params["tags"] = ",".join(tags)
        if content_type is not None:
            params["content_type"] = content_type
        return self._get_list("/knowledge-items", params=params or None, model=KnowledgeItem)

    def get(self, item_id: str) -> KnowledgeItem:
        return self._get(f"/knowledge-items/{item_id}", model=KnowledgeItem)

    def update(self, item_id: str, payload: KnowledgeItemUpdate | Mapping[str, Any]) -> KnowledgeItem:
        return self._put(f"/knowledge-items/{item_id}", payload, model=KnowledgeItem)

    def delete(self, item_id: str) -> None:
        self._delete(f"/knowledge-items/{item_id}")

    def query(self, *, tags: list[str], limit: int = 10) -> KnowledgeQueryResponse:
        return self._post(
            "/knowledge-items/query",
            {"tags": tags, "limit": limit},
            model=KnowledgeQueryResponse,
        )

    def upload(
        self,
        *,
        source_id: str,
        file: bytes | BinaryIO | str | Path,
        filename: str | None = None,
        content_type: str | None = None,
        tags: list[str] | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> KnowledgeItem:
        close_after = False
        file_handle: BinaryIO | None = None
        file_name = filename

        if isinstance(file, (str, Path)):
            path = Path(file)
            file_handle = path.open("rb")
            close_after = True
            file_name = file_name or path.name
            file_payload: bytes | BinaryIO = file_handle
        elif hasattr(file, "read"):
            file_handle = file  # type: ignore[assignment]
            file_name = file_name or getattr(file, "name", "upload.bin")
            file_payload = file_handle
        else:
            file_payload = file
            file_name = file_name or "upload.bin"

        try:
            files = {
                "file": (
                    file_name,
                    file_payload,
                    content_type or "application/octet-stream",
                )
            }
            data = {
                "source_id": source_id,
                "tags": json.dumps(tags or []),
                "metadata": json.dumps(metadata or {}),
            }
            return self._post("/knowledge-items/upload", model=KnowledgeItem, files=files, data=data)
        finally:
            if close_after and file_handle is not None:
                file_handle.close()

    def download(self, item_id: str) -> DownloadedContent:
        response = self._client.raw_request("GET", f"/knowledge-items/{item_id}/content")
        self._client.raise_for_status(response)
        return DownloadedContent(
            content=response.content,
            content_type=response.headers.get("content-type"),
            filename=parse_content_disposition_filename(
                response.headers.get("content-disposition")
            ),
        )

File: packages/flutter/lib/src/resources/knowledge_items.dart
/// Knowledge items resource — `/api/knowledge-items`.
library;

import 'dart:convert';

import 'package:http/http.dart' as http;

import '../models.dart';
import '../utils.dart';
import 'base.dart';

class KnowledgeItemsResource extends BaseResource {
  const KnowledgeItemsResource(super.client);

  Future<List<KnowledgeItem>> list({
    String? sourceId,
    List<String>? tags,
    String? contentType,
  }) async {
    final params = <String, String>{
      if (sourceId != null) 'source_id': sourceId,
      if (tags != null && tags.isNotEmpty) 'tags': tags.join(','),
      if (contentType != null) 'content_type': contentType,
    };
    final data = await client.request(
      'GET',
      'knowledge-items',
      queryParameters: params.isEmpty ? null : params,
    );
    if (data is List) {
      return data.whereType<Map<String, dynamic>>().map(KnowledgeItem.fromJson).toList();
    }
    return const [];
  }

  Future<KnowledgeItem> get(String id) async {
    final data = await client.request('GET', 'knowledge-items/$id');
    return KnowledgeItem.fromJson((data as Map<String, dynamic>?) ?? {});
  }

  Future<KnowledgeItem> create(Map<String, dynamic> payload) async {
    final data = await client.request('POST', 'knowledge-items', body: payload);
    return KnowledgeItem.fromJson((data as Map<String, dynamic>?) ?? {});
  }

  Future<KnowledgeItem> update(String id, Map<String, dynamic> payload) async {
    final data =
        await client.request('PUT', 'knowledge-items/$id', body: payload);
    return KnowledgeItem.fromJson((data as Map<String, dynamic>?) ?? {});
  }

  Future<void> delete(String id) =>
      client.request('DELETE', 'knowledge-items/$id');

  /// `POST /api/knowledge-items/query` — semantic / tag-based query.
  Future<Map<String, dynamic>> query({
    required List<String> tags,
    int limit = 10,
  }) async {
    final data = await client.request(
      'POST',
      'knowledge-items/query',
      body: {'tags': tags, 'limit': limit},
    );
    return (data as Map<String, dynamic>?) ?? {};
  }

  /// `POST /api/knowledge-items/upload` — multipart file upload.
  ///
  /// [bytes] is the raw file content.  [filename] defaults to `'upload.bin'`.
  Future<KnowledgeItem> upload({
    required String sourceId,
    required List<int> bytes,
    String filename = 'upload.bin',
    String contentType = 'application/octet-stream',
    List<String> tags = const [],
    Map<String, dynamic> metadata = const {},
  }) async {
    final file = http.MultipartFile.fromBytes(
      'file',
      bytes,
      filename: filename,
      contentType: http.MediaType.parse(contentType),
    );

    final data = await client.multipartRequest(
      'knowledge-items/upload',
      fields: {
        'source_id': sourceId,
        'tags': json.encode(tags),
        'metadata': json.encode(metadata),
      },
      files: [file],
    );
    return KnowledgeItem.fromJson((data as Map<String, dynamic>?) ?? {});
  }

  /// `GET /api/knowledge-items/{id}/content` — binary download.
  Future<DownloadedContent> download(String id) async {
    final response = await client.rawRequest('GET', 'knowledge-items/$id/content');
    if (response.statusCode >= 400) {
      throw Exception('Download failed: HTTP ${response.statusCode}');
    }
    return DownloadedContent(
      bytes: response.bodyBytes,
      contentType: response.headers['content-type'],
      filename: parseContentDispositionFilename(
        response.headers['content-disposition'],
      ),
    );
  }
}
